tests/idris2/basic032/Idiom.idr

```diff
@@ -12,4 +12,3 @@ export
 
 addm : Maybe Int -> Maybe Int -> Maybe Int
 addm x y = [| x + y |]
-
```

Comment:
If you add an example like

```idris
pairing : Applicative f => f a -> f b -> f (a, b)
pairing a b = [| (a, b) |]
```

at the bottom here, you get an ambiguity error. I wonder whether
it is solved if you make `IAlternative` and `<*>` commute like I
suggested above.

Reply:
There's more going on here than I expected at first in regards to IAlternative and how idris typechecks them.
What we (roughly) get for `pairing : Applicative f => f a -> f b -> f (a, b)` with the current fix is:
```idris
IAlternative fc
  (UniqueDefault
    (IApp fc
      (IApp fc
        (IVar fc (UN "<*>"))
        (IApp fc
          (IApp fc
            (IVar fc (UN "<*>"))
            (IApp fc
              (IVar fc (UN "pure"))
              (IVar fc (NS (MkNS ["Builtin"]) (UN "MkPair")))))
          (IVar fc (UN "x"))))
      (IVar fc (UN "y"))))
    [IApp fc
      (IApp fc
        (IVar fc (UN "<*>"))
        (IApp fc
          (IApp fc
            (IVar fc (UN "<*>"))
            (IApp fc
              (IVar fc (UN "pure"))
              (IVar fc (NS (MkNS ["Builtin"]) (UN "Pair")))))
          (IVar fc (UN "x"))))
      (IVar fc (UN "y"))
     ,IApp fc
       (IApp fc
         (IVar fc (UN "<*>"))
         (IApp fc
           (IApp fc
             (IVar fc (UN "<*>"))
             (IApp fc
               (IVar fc (UN "pure"))
                 (IVar fc (NS (MkNS ["Builtin"]) (UN "MkPair")))))
                 (IVar fc (UN "x"))))
           (IVar fc (UN "y"))]
```
And what would actually work is if instead we could produce:
```idris
(IApp fc
  (IApp fc 
    (IVar fc (UN "<*>"))
    (IApp fc
      (IApp fc
        (IVar fc (UN "<*>"))
        (IApp fc
          (IVar fc (UN "pure"))
          (IAlternative fc (UniqueDefault (IVar fc (NS (MkNS ["Builtin"]) (UN "MkPair"))))
             [(IVar fc (NS (MkNS ["Builtin"]) (UN "MkPair"))), (IVar fc (NS (MkNS ["Builtin"]) (UN "Pair")))])))
      (IVar fc (UN "x"))))
  (IVar fc (UN "y")))
```
It's unclear to me why the first version is ambiguous when the type of `pairing` should inform us we're after MkPair where we're using `[| (a, b) |]`. The second version was accepted as the body for `pairing` via feeding it to `%runElab check` however.

I wasn't quite sure how to write your given code to see if it helped because it's just a little off, `app` in particular, but I tried some variations without luck.

Reply:
I have tried to do something like this:

```diff
+idiomisePure : FC -> RawImp -> RawImp
+idiomisePure fc fn = IApp fc (IVar fc (UN "pure")) fn
+
 idiomise : FC -> RawImp -> RawImp
-idiomise fc (IAlternative afc u alts)
-  = IAlternative afc (mapAltType (idiomise afc) u) (idiomise afc <$> alts)
+idiomise fc fn@(IAlternative afc u alts)
+  = if any (isJust . isIApp) alts
+   -- if we have at least an `App` then we cannot do any better than idiomising
+   -- under the alternative
+    then IAlternative afc (mapAltType (idiomise afc) u) (idiomise afc <$> alts)
+    -- if there is no `App` however we know all of these will be caught by the
+    -- pure case and we can put the pure layer on top of the `IAlternative` one.
+    -- This means the applicative can get disambiguated first, and the other ambiguity
+    -- second!
+    else idiomisePure fc fn
 idiomise fc (IApp afc f a)
     = IApp fc (IApp fc (IVar fc (UN "<*>"))
                     (idiomise afc f))
-                    a
-idiomise fc fn = IApp fc (IVar fc (UN "pure")) fn
+              a
+idiomise fc fn = idiomisePure fc fn
```

and then desugaring a `PPair` to (an `IAlternative` of the two constructors) applied
to the arguments rather than an `IAlternative` of (the two constructors applied to the
arguments):

```diff
   desugarB side ps (PPair fc l r)
       = do l' <- desugarB side ps l
            r' <- desugarB side ps r
-           let pval = apply (IVar fc mkpairname) [l', r']
-           pure $ IAlternative fc (UniqueDefault pval)
-                  [apply (IVar fc pairname) [l', r'], pval]
+           let def = IVar fc mkpairname
+           pure $ apply (IAlternative fc (UniqueDefault def) [IVar fc pairname, def])
+                        [l', r']
```

However this last change breaks *a lot* of things as we pick an `IAlternative`
based on its content only & then commit to it. So `IAlternative` between two
vars will always succeed with the first one AFAICT!

I don't really see a good way to fix this problem so that `pairing` works.
I would be in favour of merging as-is as it's already an improvement on the
current state.